dropout1d [#5105], Extend Ivy Functional API [#3856]

MuhammadSaeedBatikh opened the pull request:
Maybe it's a good idea to create a unified dropoutxd function and have dropout1d, dropout2d, and dropout3d perform dimensionality checking and call dropoutxd instead of writing separate, identical logic for each?

CatB1t reviewed (changes requested):
Hello @MuhammadSaeedBatikh, Thanks for the great work! I'll be reaching out to someone from the team regarding your suggestion, as this is something we're thinking about also for `convd`.

MuhammadSaeedBatikh commented on ivy/functional/backends/jax/layers.py:
```diff
@@ -1,14 +1,19 @@
+# Extra #
+# ------#
+
+RNG = jax.random.PRNGKey(0)
```
Hello @CatB1t, thanks for your review. It's now using `RNG` key from `jax/random.py`.

MuhammadSaeedBatikh commented:
Hi @ivy-leaves, and @CatB1t.

I think some of the failing tests might not be related to the task!
For example, the first check: `test-docstrings / run-docstring-tests`
the log says that there's an error at module `ivy/container/extensions.py` in method ``static_moveaxis``
caused by a missing blank line after section ('Parameters') and ('Returns'). I do not believe this is related to dropout1d since all of its container updates are in `container/layers.py` and not `container/extensions.py`.

For the second check: `test-numpy-style-docstrings / run-docstring-tests`
it says "The following methods had failing docstrings:"  and lists methods such as `Container.abs`, ``Container.layer_norm``, Container.conv1d`, and so forth; however, it does not list dropout1d.

MuhammadSaeedBatikh commented:
There are also other checks that are failing for some backends and working for others.
For example, `linalg` appears to be working only for numpy. The log indicates that `test_matrix_transpose`, `test_outer `, and `test_trace` have failed.
Also, `manipulation_functions` is working for numpy and jax, but failing for torch and tensorflow. The log for both failed backends outputs **test_squeeze FAILED**.

I am not sure how this is related to dropout1d. Any ideas?

CatB1t commented on ivy_tests/test_ivy/test_functional/test_nn/test_layers.py:
```diff
@@ -140,6 +140,60 @@ def test_dropout(
+        x=x[0],
+        prob=prob,
+        training=training,
+        dtype=dtype,
```
the function doesn't take a `dtype` parameter

CatB1t commented on ivy/functional/backends/torch/layers.py:
```diff
@@ -674,3 +674,24 @@ def conv_general_transpose(
+    if data_format == "NWC":
+        perm = (0, 2, 1) if len(x.shape) == 3 else (1, 0)
+        x = torch.permute(x, perm)
+    res = torch.nn.functional.dropout1d(x, p=prob, training=training)
```
our torch version is pinned to an older release, which doesn't have `dropout1d`, running the test locally throws an exception, you should implement it manually for now, adding a # ToDo to be updated with torch version is updated.

CatB1t commented on ivy_tests/test_ivy/test_functional/test_nn/test_layers.py:
```diff
@@ -140,6 +140,60 @@ def test_dropout(
+        training=training,
+        dtype=dtype,
+    )
+    ret = helpers.flatten_and_to_np(ret=ret)
```
you can use `return_flat_np_arrays` flag instead of manually doing it

CatB1t commented on ivy_tests/test_ivy/test_functional/test_nn/test_layers.py:
```diff
@@ -140,6 +140,60 @@ def test_dropout(
+    device,
+):
+    dtype, x = dtype_and_x
+    ret = helpers.test_function(
```
it's better if we also test that the result also match the `ground_truth`

CatB1t reviewed (changes requested):
Please, make sure you run the test locally again with `training=True` after removing the `dtype` parameter, that was causing the test to silently pass, the tests seems to be failing for me locally for all backends.

CatB1t commented:
> Hi @ivy-leaves, and @CatB1t.
> 
> I think some of the failing tests might not be related to the task! For example, the first check: `test-docstrings / run-docstring-tests` the log says that there's an error at module `ivy/container/extensions.py` in method `static_moveaxis` caused by a missing blank line after section ('Parameters') and ('Returns'). I do not believe this is related to dropout1d since all of its container updates are in `container/layers.py` and not `container/extensions.py`.
> 
> For the second check: `test-numpy-style-docstrings / run-docstring-tests` it says "The following methods had failing docstrings:" and lists methods such as `Container.abs`, `Container.layer_norm`, Container.conv1d`, and so forth; however, it does not list dropout1d.



> There are also other checks that are failing for some backends and working for others. For example, `linalg` appears to be working only for numpy. The log indicates that `test_matrix_transpose`, `test_outer `, and `test_trace` have failed. Also, `manipulation_functions` is working for numpy and jax, but failing for torch and tensorflow. The log for both failed backends outputs **test_squeeze FAILED**.
> 
> I am not sure how this is related to dropout1d. Any ideas?

You can ignore these if they're not related to your task.

MuhammadSaeedBatikh commented:
Hi @CatB1t, thanks for the very detailed review!
Regarding the requested changes; I added an additional test for `data_format`, reimplemented `dropout1d` from scratch for older torch versions, and fixed the tests that were failing. 
There are 9 failing checks. I believe 8 of them are unrelated to this task, however, the `run-nightly-tests (jax, layers)` seems to be relevant. 
The log for this Jax check indicates that are two distinct failures, the important one is that there is a type error when calling the `jnp.transpose` method from `jax.layers.dropout1d`.
Specifically it raises the following exception:
`TypeError: transpose requires ndarray or scalar arguments, got <class 'ivy.array.array.Array'> at position 0` 
It seems that during test, an ivy array is being passed to `jax.dropout1d` directly without conversion and subsequently passed to `jnp.transpose` , but I am not sure why/how this is happening.
Any Ideas?

MuhammadSaeedBatikh commented:
> Hi @CatB1t, thanks for the very detailed review! Regarding the requested changes; I added an additional test for `data_format`, reimplemented `dropout1d` from scratch for older torch versions, and fixed the tests that were failing. There are 9 failing checks. I believe 8 of them are unrelated to this task, however, the `run-nightly-tests (jax, layers)` seems to be relevant. The log for this Jax check indicates that are two distinct failures, the important one is that there is a type error when calling the `jnp.transpose` method from `jax.layers.dropout1d`. Specifically it raises the following exception: `TypeError: transpose requires ndarray or scalar arguments, got <class 'ivy.array.array.Array'> at position 0` It seems that during test, an ivy array is being passed to `jax.dropout1d` directly without conversion and subsequently passed to `jnp.transpose` , but I am not sure why/how this is happening. Any Ideas?

Adding `@to_native_arrays_and_back` decorator seems to resolve the issue.

CatB1t commented on ivy_tests/test_ivy/test_functional/test_nn/test_layers.py:
```diff
@@ -132,7 +132,63 @@ def test_dropout(
+        ground_truth_backend='tensorflow'
     )
     ret = helpers.flatten_and_to_np(ret=ret)
     for u in ret:
```
This should be `for u, v, k in zip(ret, gt_ret, x):`, also we should assert the shape across all 3.

CatB1t reviewed (changes requested):
Just few changes! and then we're good to go! 😅

ivy-seed commented:
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

MuhammadSaeedBatikh commented:
Hi @CatB1t, and @ivy-seed 
I think I have made the necessary changes, could you please review them?  Thanks.

CatB1t reviewed (approved):
lgtm! thanks for your contribution! 😄
